### PR TITLE
Add support and tests for nested objects in formURLEncoded parser

### PR DESCRIPTION
**Pull request by gwynne**
This fixes a bug where attempting to use a nested object would result in garbage data.

Note: The code that handles key paths in the parser is pretty gnarly, would appreciate more eyes on it. @tanner0101

### CODE REVIEW

File: Sources/Vapor/FormURLEncoded/StructuredData+FormURLEncoded.swift [hunk 1]
```diff
@@ -48,42 +48,66 @@ extension Node {
...
             if let existing = urlEncoded[key] {
                 if keyIndicatedArray {
-                    var array = existing.array ?? [existing]
-                    array.append(value)
-                    value = .array(array)
+                    if subKey.count > 0 { // there's a keypath before the array
+                        // This is an ugly mess. There must be a better way.
+                        var obj = existing
+                        if let _ = obj.object {} else { obj = Node.object([:]) }
+                        var array = obj[subKey]?.array ?? (obj[subKey] != nil ? [obj[subKey]!] : Array<Node>())
```
Comment (tanner0101) [inline Sources/Vapor/FormURLEncoded/StructuredData+FormURLEncoded.swift]: can we use `if/let` instead of `!`

**Comment by gwynne** [comment]
@tanner0101 Good catch, fixed!

**Comment by tanner0101** [comment]
@Joannis can you take a look at this one and see if it all looks good. the code is quite complex so i'd like at least one more pair of eyes before merging. thanks!

File: Sources/Vapor/FormURLEncoded/StructuredData+FormURLEncoded.swift [hunk 2]
```diff
@@ -48,42 +48,68 @@ extension Node {
...
-                if contents[0] == .rightSquareBracket {
-                    keyIndicatedArray = true
-                } else {
-                    subKey = contents.dropLast().makeString()
-                    keyIndicatedObject = true
+                
+                for (n, contents) in slices.dropFirst().enumerated() {
+                    let content = contents.array
+                    
+                    if content[0] == .rightSquareBracket {
```
Comment (Joannis) [inline Sources/Vapor/FormURLEncoded/StructuredData+FormURLEncoded.swift]: The assumption that the contents have `>= 0` elements can crash the server.
I'd do a `content.first == .rightSquareBracket`. `.first` will leave it `nil` if there is no first element

File: Sources/Vapor/FormURLEncoded/StructuredData+FormURLEncoded.swift [hunk 3]
```diff
@@ -48,42 +48,68 @@ extension Node {
...
+                        // Current subsequence indicates an array.
+                        guard n == slices.endIndex - 2 else { // subtract 2 because we also ignore element 0
```
Comment (Joannis) [inline Sources/Vapor/FormURLEncoded/StructuredData+FormURLEncoded.swift]: Magic numbers are adding confusion. Could we do an index based approach here?

```swift
nIndex < slices.endIndex
```

If we check for `<` we're not accepting `>= endIndex`, keeping this non-crashing and more readable.
Also; why does `n` need to be equal to this exact number to indicate it's an array?

File: Sources/Vapor/FormURLEncoded/StructuredData+FormURLEncoded.swift [hunk 4]
```diff
@@ -48,42 +48,68 @@ extension Node {
...
+                            print("Array stem must be last in nested encoded key specifier in \(pair.makeString()) ... continuing")
```
Comment (Joannis) [inline Sources/Vapor/FormURLEncoded/StructuredData+FormURLEncoded.swift]: `print`

File: Sources/Vapor/FormURLEncoded/StructuredData+FormURLEncoded.swift [hunk 5]
```diff
@@ -48,42 +48,68 @@ extension Node {
...
+                            continue perKeyLoop
+                        }
+                        keyIndicatedArray = true
+                    } else {
+                        // Current subsequence indicates an object.
+                        // this can result in both keyIndicatedArray and keyIndicatedObject
+                        // being true. That's okay as long as array is checked first below.
+                        subKey += (subKey.characters.count > 0 ? "." : "") + contents.dropLast().makeString()
```
Comment (Joannis) [inline Sources/Vapor/FormURLEncoded/StructuredData+FormURLEncoded.swift]: `dropLast` crashes if `contents.count == 0`

File: Sources/Vapor/FormURLEncoded/StructuredData+FormURLEncoded.swift [hunk 6]
```diff
@@ -48,42 +48,68 @@ extension Node {
...
 
             let key = keyData.makeString()
 
             if let existing = urlEncoded[key] {
                 if keyIndicatedArray {
-                    var array = existing.array ?? [existing]
-                    array.append(value)
-                    value = .array(array)
+                    if subKey.characters.count > 0 { // there's a keypath before the array
+                        // This is an ugly mess. There must be a better way.
```
Comment (Joannis) [inline Sources/Vapor/FormURLEncoded/StructuredData+FormURLEncoded.swift]: I think you're right, there should be a better way. Maybe adding more whitespace and comments? Comments help more in readability than in information.

File: Tests/VaporTests/ContentTests.swift [hunk 7]
```diff
@@ -111,6 +113,31 @@ class ContentTests: XCTestCase {
         XCTAssertEqual(data["obj.foo"], "bar")
     }
 
+    func testFormURLEncodedNestedDict() {
+        let body="datatable[pagination][total]=1044&datatable[pagination][page]=1&datatable[pagination][perpage]=10&datatable[pagination][field]=name&datatable[pagination][sort]=asc&datatable[pagination][pages]=105&datatable[sort][field]=name&datatable[sort][sort]=asc&datatable[query][test]=test"
+        let data = Node(formURLEncoded: body.makeBytes(), allowEmptyValues: true)
+        let foo = try! data.converted(to: JSON.self).makeBytes().makeString()
+        print(foo)
```
[the remark this answers is not in the document]
Comment (gwynne) [reply Tests/VaporTests/ContentTests.swift]: That was part of my copy-paste from other tests - should I remove it from other tests that do this while I'm at it?

**Comment by vzsg** [comment]
I couldn't discern where these nested requests (i.e. the one in the test case) are originating from, based on the Slack discussion; but it exists in the wild, and anything works better than the broken parser.

AFAICT this is a non-standard addition on top of the key-value pairs defined in the RFC, yet some big names (for example: Stripe) use nested objects in form-urlencoded in their API. (Although they use JSON in their responses, so it's not coming from them.)

**Comment by gwynne** [comment]
Updated to address all of @Joannis ' comments, thanks @Joannis for the review!

@vzsg I don't believe the nesting support is defined by any standard, but as you pointed out, others support it, and I don't think there's any reason for us not to.

**Comment by Joannis** [comment]
Really solid update in that last commit 👍 I'm fine here with my code quality review.

I agree with @vzsg , though. The additions are really clean now, but the scope of this function should stay clear.

If we want additional functionality in the same spec, it's probably best to extend the functionality in a separate function or some other route.

I'll leave the rest of the discussion to @vzsg and @tanner0101

**Comment by gwynne** [comment]
My own main concern regarding functionality here is that this adds it for `application/x-www-form-urlencoded` inputs, but `multipart/form-data` submissions do not get the same support (adding it wouldn't be _too_ difficult, just some tweaks in the `onField` handler in `Request+FormData.swift`, but whether to do so is a weird question. Currently that parser doesn't even support single-level arrays or objects.

**Comment by gwynne** [comment]
Obsoleted by Vapor 3.

**Pull request closed without merging**